client/src/main/java/org/apache/rocketmq/client/impl/consumer/DefaultMQRealPushConsumerImpl.java

```diff
@@ -0,0 +1,1305 @@
+            return false;
+        } else {
+            //Stop pull request
+            log.debug("#####Process Push : Current Local key:{} and  offset:{} and push offset:{}", localOffsetKey, localOffset.get(), offset);
+            pullStop = this.pullStopped.get(localOffsetKey);
+            if (!pullStop.get()) {
+                pullStop.set(true);
+                log.info("Pull stop of {} is set to TRUE, and then the pull request will stop...", localOffsetKey);
+            }
+            //update local offset
+            localOffset.set(offset);
+            //submit to process queue
+            List<MessageExt> messageExtList = new ArrayList<MessageExt>();
+            messageExtList.add(msg);
+
+            processQueue.putMessage(messageExtList);
+            if (this.consumeOrderly) {
+                processQueue.setLocked(true);
+                processQueue.setLastLockTimestamp(System.currentTimeMillis());
+            }
+            MessageQueue messageQueue = new MessageQueue(topic, brokerName, queueID);
+            //update process Queue pull time
+            if (this.rebalanceImpl.processQueueTable.get(messageQueue) != null) {
+                this.rebalanceImpl.processQueueTable.get(messageQueue).setLastPullTimestamp(System.currentTimeMillis());
+            }
+            this.consumeMessageService.submitConsumeRequest(messageExtList, processQueue, messageQueue, true);
+        }
+        return true;
+    }
+
+    private String genLocalOffsetKey(String consumerGroup, String topic, String brokerName, int queueID) {
+        return consumerGroup + "@" + topic + "@" + brokerName + "@" + queueID;
+    }
+
+    public boolean resumePullRequest(String consumerGroup, String topic, String brokerName, int queueID) {
+        String localOffsetKey = genLocalOffsetKey(consumerGroup, topic, brokerName, queueID);
+        ProcessQueue processQueue = processQueues.get(localOffsetKey);
+        if (processQueue != null) {
+            log.info("Clear local expire message for {} in processQueue.", localOffsetKey);
+            processQueue.cleanExpiredMsg(this.defaultMQPushConsumer);
+        }
+        AtomicBoolean pullStop = this.pullStopped.get(localOffsetKey);
+        if (pullStop != null) {
+            if (pullStop.get()) {
+                pullStop.set(false);
+                log.info("Resume Pull Request of {} is set to TRUE, and then the pull request will start by rebalance again...", localOffsetKey);
+            }
+        }
+        return true;
+    }
+
+    public boolean pausePullRequest(String consumerGroup, String topic, String brokerName, int queueID) {
+        String localOffsetKey = genLocalOffsetKey(consumerGroup, topic, brokerName, queueID);
+        AtomicBoolean pullStop = this.pullStopped.get(localOffsetKey);
+        if (pullStop == null) {
+            this.pullStopped.put(localOffsetKey, new AtomicBoolean(true));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `get()`, `null` check, and `put()` may not be thread-safe at lines: **1293, 1294, and 1295**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread's `putIfAbsent()`.

**More info**  
 `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.

**More info**  
 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/startupprogress/StartupProgress.java#L272-L273) (external link).